src/compiler/checker.ts

```diff
@@ -23935,9 +23935,21 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     function inferReverseMappedType(sourceType: Type, target: MappedType, constraint: IndexType): Type {
         const typeParameter = getIndexedAccessType(constraint.type, getTypeParameterFromMappedType(target)) as TypeParameter;
         const templateType = getTemplateTypeFromMappedType(target);
-        const inference = createInferenceInfo(typeParameter);
-        inferTypes([inference], sourceType, templateType);
-        return getTypeFromInference(inference) || unknownType;
+        const inferences = [createInferenceInfo(typeParameter)];
+        inferTypes(inferences, sourceType, templateType);
+        const wholeInferredType = getTypeFromInference(inferences[0]);
+        if (!wholeInferredType && inferences.length > 1) {
+            const members = createSymbolTable();
+            for (let i = 1; i < inferences.length; i++) {

```

Comment:
This assumes that we only collect inferences for **direct** properties of the reverse mapped type. I think that this is enough for a lot of cases but we could also try to infer deeper ones. This would complicate the algorithm a little bit though. Right now I can just "merge" all of those extra inferences into a single object. If we'd infer into nested properties then:
- this merging would have to be smart enough to handle properties at arbitrary levels when those inferences are kept in a flat array and they don't have any particular order that would help the merging algorithm. Maybe the layout of those extra inferences could be improved to aid this merging algorithm but then that would complicate the `inferences.push(inference)` call site
- we should decide what to do when we find inferences for both `T[K]['foo']` and `T[K]['foo']['a']`. That is pretty much the same situation as `T[K]` and `T[K]['foo']`, so I think that we simply should always "prefer" the less nested one

Reply:
There is also an open question if this whole feature is a good idea. One could argue that this is inconsistent with other parts of TS since TS never collects such "partial"/prop-based inferences elsewhere... OTOH, it actually kinda does this with `T[K]` in the reverse mapped type.

So I treat this feature as an extension of the existing mechanism for reverse mapped types. I think that this would enable some great new capabilities for library authors because this feature would allow inferring **multiple** unique things **per** tuple element/object property. Right now it's only possible to infer a **single** unique thing **per** tuple element/object property and that is a little bit limiting when it comes to expressing complex types.

Reply:
> This assumes that we only collect inferences for direct properties of the reverse mapped type.

while the implementation has completely changed (basically I pulled https://github.com/microsoft/TypeScript/pull/20126 here and just tweaked it a little bit) - this is still a "limitation" but I still believe that it's fine. So this doesn't work with the current state of this PR:
```ts
declare function test3<
  T extends { p1: { p1p1: unknown; p1p2: unknown }; p2: unknown }
>(o: { a: T["p1"]["p1p1"]; b: T["p2"]; c: T["p1"]["p1p2"] }): T;

test3({
  a: "foo",
  b: 42,
  c: true,
});
```

> There is also an open question if this whole feature is a good idea. One could argue that this is inconsistent with other parts of TS since TS never collects such "partial"/prop-based inferences elsewhere... 

This is no longer the case with this PR since this infers now:
```ts
declare function test2<T extends { p1: unknown; p2: unknown }>(o: {
  a: T["p1"];
  b: T["p2"];
}): T;

// `T` inferred as `{ p1: string; } & { p2: number; }`
test2({
  a: "foo",
  b: 42,
});
```